File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/GpxWindow.java
package ch.bailu.aat_lib.gpx;

import javax.annotation.Nonnull;

import ch.bailu.aat_lib.coordinates.BoundingBoxE6;
import ch.bailu.aat_lib.gpx.interfaces.GpxDeltaInterface;

public abstract class GpxWindow implements GpxDeltaInterface {

    private GpxPointNode first;
    private GpxPointNode last;

    private float distance;
    private long timeDeltaMillis;

    public GpxWindow(GpxPointNode node) {
        first = node;
        last = node;
    }

    public void forward(GpxPointNode n) {
        last = n;

        distance += n.getDistance();
        timeDeltaMillis += n.getTimeDelta();

        trim();
    }

    private void trim() {
        while (overLimit() && first != last && first.getNext() instanceof GpxPointNode) {
            timeDeltaMillis -= first.getTimeDelta();
            distance -= first.getDistance();

            first = (GpxPointNode) first.getNext();
        }
    }

    protected abstract boolean overLimit();

    @Override
    public float getDistance() {
        return distance;
    }


    @Override
    public float getSpeed() {
        if (timeDeltaMillis > 0) {
            return (distance * 1000f) / timeDeltaMillis;
        } else {
            return 0;
        }
    }

    @Override
    public float getAcceleration() {
        return 0;
    }

    @Override
    public long getTimeDelta() {
        return timeDeltaMillis;
    }

    @Nonnull
    @Override
    public BoundingBoxE6 getBoundingBox() {
        return BoundingBoxE6.NULL_BOX;
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/GpxWindow.kt
package ch.bailu.aat_lib.gpx

import ch.bailu.aat_lib.coordinates.BoundingBoxE6
import ch.bailu.aat_lib.gpx.interfaces.GpxDeltaInterface
import javax.annotation.Nonnull

/** Sliding window over a linked list of {@link GpxPointNode}s, trimming from the front when a subclass-defined limit is exceeded. */
abstract class GpxWindow(private var first: GpxPointNode?) : GpxDeltaInterface {
    private var last: GpxPointNode?

    private var distance = 0f
    private var timeDeltaMillis: Long = 0

    init {
        last = first
    }

    fun forward(node: GpxPointNode) {
        last = node

        distance += node.getDistance()
        timeDeltaMillis += node.getTimeDelta()

        trim()
    }

    private fun trim() {
        while (overLimit() && first !== last && first!!.next is GpxPointNode) {
            timeDeltaMillis -= first!!.getTimeDelta()
            distance -= first!!.getDistance()

            first = first!!.next as GpxPointNode?
        }
    }

    protected abstract fun overLimit(): Boolean

    override fun getDistance(): Float {
        return distance
    }


    override fun getSpeed(): Float {
        if (timeDeltaMillis > 0) {
            return (distance * 1000f) / timeDeltaMillis
        } else {
            return 0f
        }
    }

    override fun getAcceleration(): Float {
        return 0f
    }

    override fun getTimeDelta(): Long {
        return timeDeltaMillis
    }

    @Nonnull
    override fun getBoundingBox(): BoundingBoxE6 {
        return BoundingBoxE6.NULL_BOX
    }
}
